SharpCaster/ChromeCastClient.cs

```diff
@@ -1,222 +1,330 @@
                 {
-                    var buffer = new byte[stream.Length];
-                    await stream.ReadAsync(buffer, 0, buffer.Length, cancellationToken);
-                    entireMessage = buffer;
+                    disposable.Dispose();
+                }
+                catch (Exception ex) {
+                    Logger.InfoException("Error on disposing.", ex, null);
                 }
-                else
+                finally
                 {
-                    entireMessage = await stream.ParseData(cancellationToken);
+                    action();
                 }
-
-                var entireMessageArray = entireMessage.ToArray();
-                var castMessage = entireMessageArray.ToCastMessage();
-                if (string.IsNullOrEmpty(castMessage?.Namespace)) return;
-
-                Debug.WriteLine("Received: " + castMessage.GetJsonType());
-                ReceivedMessage(castMessage);
             }
-            catch (Exception ex)
-            {
-                // TODO: Catch disconnect - HResult = 0x80072745 -
-                // catch this (remote device disconnect) ex = {"An established connection was aborted
-                // by the software in your host machine. (Exception from HRESULT: 0x80072745)"}
+        }
 
-                // Log these bytes
-                Debug.WriteLine(ex);
-            }
+        /// <summary>
+        /// Raises the Disconnected event
+        /// </summary>
+        protected virtual void OnDisconnected()
+        {
+            Disconnected?.Invoke(this, EventArgs.Empty);
+        }
+        
+        /// <summary>
+        /// Gets a channel
+        /// </summary>
+        /// <typeparam name="TChannel">channel type</typeparam>
+        /// <returns>a channel</returns>
+        public TChannel GetChannel<TChannel>() where TChannel : IChromecastChannel
+        {
+            return Channels.OfType<TChannel>().FirstOrDefault();
         }
 
-        private void ReceivedMessage(CastMessage castMessage)
+        public async Task<ChromecastStatus> LaunchApplicationAsync(string applicationId, bool joinExistingApplicationSession = true)
         {
-            foreach (var channel in Channels.Where(i => i.Namespace == castMessage.Namespace))
+            if (joinExistingApplicationSession)
             {
-                channel.OnMessageReceived(new ChromecastSSLClientDataReceivedArgs(castMessage));
+                var status = GetChromecastStatus();
+                var runningApplication = status?.Applications?.FirstOrDefault(x => x.AppId == applicationId);
+                if (runningApplication != null)
+                {
+                    await GetChannel<IConnectionChannel>().ConnectAsync(runningApplication.TransportId);
+                    return await GetChannel<IReceiverChannel>().GetChromecastStatusAsync();
+                }
             }
+            return await GetChannel<IReceiverChannel>().LaunchApplicationAsync(applicationId);
```

Comment:
I used your code and I found that this method only gets me a working ChromeCastClient if it takes the 'rejoin path'. If it creates a new Application with the LaunchApplicationAsync on the receiverChannel the Application is started but the media channel is not working.
I could fix this by executing the same 2 lines you use to (re)join the application in case there is one after creating the new one. I am not sure this is the best way to sove it but it works for me. See https://github.com/RobertK66/SharpCaster/commit/0759dbc6b94947933024b93852354aeb24bde48e for more details...